beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs

```diff
@@ -0,0 +1,331 @@
+    db: Arc<HotColdDB<T::EthSpec, T::HotStore, T::ColdStore>>,
+) -> Result<(), StoreError> {
+    // `PersistedForkChoice` stores the `ProtoArray` as a `Vec<u8>`. Deserialize these
+    // bytes assuming the legacy struct, and transform them to the new struct before
+    // re-serializing.
+    let ssz_container_v6 =
+        SszContainerV6::from_ssz_bytes(&persisted_fork_choice.fork_choice.proto_array_bytes)
+            .map_err(|e| {
+                StoreError::SchemaMigrationError(format!(
+                    "Failed to decode ProtoArrayForkChoice during schema migration: {:?}",
+                    e
+                ))
+            })?;
+
+    // Clone the V6 proto nodes in order to maintain information about `node.justified_epoch`
+    // and `node.finalized_epoch`.
+    let nodes_v6 = ssz_container_v6.nodes.clone();
+
+    let justified_checkpoint = persisted_fork_choice.fork_choice_store.justified_checkpoint;
+    let finalized_checkpoint = persisted_fork_choice.fork_choice_store.finalized_checkpoint;
+
+    // These transformations instantiate `node.justified_checkpoint` and `node.finalized_checkpoint`
+    // to `None`.
+    let ssz_container_v7: SszContainerV7 =
+        ssz_container_v6.into_ssz_container_v7(justified_checkpoint, finalized_checkpoint);
+    let ssz_container: SszContainer = ssz_container_v7.into();
+    let mut fork_choice: ProtoArrayForkChoice = ssz_container.into();
+
+    update_checkpoints::<T>(finalized_checkpoint.root, &nodes_v6, &mut fork_choice, db)
+        .map_err(StoreError::SchemaMigrationError)?;
+
+    // Update the justified checkpoint in the store in case we have a discrepancy
+    // between the store and the proto array nodes.
+    update_store_justified_checkpoint(persisted_fork_choice, &mut fork_choice)
+        .map_err(StoreError::SchemaMigrationError)?;
+
+    Ok(())
+}
+
+struct HeadInfo {
+    index: usize,
+    root: Hash256,
+    slot: Slot,
+}
+
+fn update_checkpoints<T: BeaconChainTypes>(
+    finalized_root: Hash256,
+    nodes_v6: &[ProtoNodeV6],
+    fork_choice: &mut ProtoArrayForkChoice,
+    db: Arc<HotColdDB<T::EthSpec, T::HotStore, T::ColdStore>>,
+) -> Result<(), String> {
+    let heads = find_finalized_descendant_heads(finalized_root, fork_choice);
+
+    // For each head, first gather all epochs we will need to find justified or finalized roots for.
+    for head in heads {
+        // `relevant_epochs` are epochs for which we will need to find the root at the start slot.
+        // We don't need to worry about whether the are finalized or justified epochs.
+        let mut relevant_epochs = vec![];
+        let relevant_epoch_finder = |index, _: &mut ProtoNode| {
+            let (justified_epoch, finalized_epoch) = nodes_v6
+                .get(index)
+                .map(|node: &ProtoNodeV6| (node.justified_epoch, node.finalized_epoch))
+                .ok_or_else(|| "Head index not found in legacy proto nodes".to_string())?;
```

Comment:
```suggestion
                .ok_or_else(|| "Index not found in legacy proto nodes".to_string())?;
```

Sometimes this function will be supplied indices of ancestors of the head.